### PR TITLE
Optimized pre-pull scan for conflicted docs

### PR DESCRIPTION
**Pull request by snej**
C4DocEnumerator:
* Added option flag for unordered results.
* Conflicts-only option is passed through to the RecordEnumerator.
* Stripped out unused/unneeded functionality, like filtering.

RecordEnumerator:
* Added options for conflicts-only & unordered results.
* With Debug-level query logging, it explains its query plan.

Database:
* Added SQLite partial indexes for conflicted docs and docs with blobs.
* During compaction, enumeration of docs with blobs is now unordered.

Replicator:
* The conflicted-docs scan is unordered, to speed it up.
* Logging before and after the scan, including the time taken.

Fixes [CBL-536](https://issues.couchbase.com/browse/CBL-536) (but hasn't been benchmarked on large dbs!)

### CODE REVIEW

File: C/c4DocEnumerator.cc [hunk 1]
```diff
@@ -36,89 +35,51 @@ CBL_CORE_API const C4EnumeratorOptions kC4DefaultEnumeratorOptions = {
...
-        do {
-            if (!_e.next())
-                return false;
-        } while (!useDoc());
-        return true;
-    }
-
     Retained<Document> getDoc() {
-        return _e ? _database->documentFactory().newDocumentInstance(_e.record()) : nullptr;
+        return *this ? _database->documentFactory().newDocumentInstance(record()) : nullptr;
```
Comment (borrrden) [inline C/c4DocEnumerator.cc]: boolean test for `*this`?  Is there any way to be clearer about what this is actually checking?

File: LiteCore/Storage/RecordEnumerator.cc [hunk 2]
```diff
@@ -82,7 +74,7 @@ namespace litecore {
                 close();
                 return false;
             }
-            LogToAt(EnumLog, Debug, "enum:     --> [%s]", _record.key().hexCString());
+            //LogToAt(QueryLog, Debug, "RecordEnumerator %p  --> '%.*s'", this, SPLAT(_record.key()));
```
Comment (borrrden) [inline LiteCore/Storage/RecordEnumerator.cc]: Is there a reason for leaving this in as a comment?

File: C/c4DocEnumerator.cc (hunk 1 above)
Comment (snej) [reply C/c4DocEnumerator.cc]: I know, it's kind of ugly. The class has an `operator bool` to test for EOF, but not a named method. I'll add a named method so it's clearer.

File: LiteCore/Storage/RecordEnumerator.cc (hunk 2 above)
Comment (snej) [reply LiteCore/Storage/RecordEnumerator.cc]: It's very useful in specific circumstances, when debugging RecordEnumerator as I was yesterday, but it spews an AWFUL LOT of stuff during unit tests. I guess Debug logging isn't going to be on often, though, so I could leave it there.

**Pull request merged**